core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/EnvelopeWrappedSdkPublisher.java

```diff
@@ -0,0 +1,151 @@
+    private final T contentSuffix;
+    private final BiFunction<T, T, T> mergeContentFunction;
+
+    private EnvelopeWrappedSdkPublisher(Publisher<T> wrappedPublisher,
+                                        T contentPrefix,
+                                        T contentSuffix,
+                                        BiFunction<T, T, T> mergeContentFunction) {
+        this.wrappedPublisher = wrappedPublisher;
+        this.contentPrefix = contentPrefix;
+        this.contentSuffix = contentSuffix;
+        this.mergeContentFunction = mergeContentFunction;
+    }
+
+    /**
+     * Create a new publisher that wraps the content of an existing publisher.
+     * @param wrappedPublisher The publisher who's content will be wrapped.
+     * @param contentPrefix The content to be inserted in front of the wrapped content.
+     * @param contentSuffix The content to be inserted at the back of the wrapped content.
+     * @param mergeContentFunction A function that will be used to merge the inserted content into the wrapped content.
+     * @param <T> The content type.
+     * @return A newly initialized instance of this class.
+     */
+    public static <T> EnvelopeWrappedSdkPublisher<T> of(Publisher<T> wrappedPublisher,
+                                                        T contentPrefix,
+                                                        T contentSuffix,
+                                                        BiFunction<T, T, T> mergeContentFunction) {
+        return new EnvelopeWrappedSdkPublisher<>(wrappedPublisher, contentPrefix, contentSuffix, mergeContentFunction);
+    }
+
+    /**
+     * See {@link Publisher#subscribe(Subscriber)}
+     */
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        wrappedPublisher.subscribe(new ContentWrappedSubscriber(subscriber));
+    }
+
+    private class ContentWrappedSubscriber implements Subscriber<T> {
+        private final Subscriber<? super T> wrappedSubscriber;
+        private final AtomicBoolean prefixApplied = new AtomicBoolean(false);
+        private final AtomicBoolean suffixApplied = new AtomicBoolean(false);
+
+        private ContentWrappedSubscriber(Subscriber<? super T> wrappedSubscriber) {
+            this.wrappedSubscriber = wrappedSubscriber;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            wrappedSubscriber.onSubscribe(subscription);
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (contentPrefix != null && prefixApplied.compareAndSet(false, true)) {
+                wrappedSubscriber.onNext(mergeContentFunction.apply(contentPrefix, t));
+            } else {
+                wrappedSubscriber.onNext(t);
+            }
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            wrappedSubscriber.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
```

Comment:
I'm not sure, but I think this behavior might violate the spec. A `Publisher` is allowed to signal `onComplete` without first getting a `request()` signal, so it's possible that this calls `onNext` before the downstream subscriber signals any demand which is not allowed

Reply:
As discussed we'll change this to not send the envelope at all on an empty stream which is the only scenario we think this will crop up in.